// @serenity-js/core/lib/errors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnknownError = exports.TimeoutExpiredError = exports.TestCompromisedError = exports.RuntimeError = exports.LogicError = exports.ListItemNotFoundError = exports.ImplementationPendingError = exports.ErrorStackParser = exports.ErrorSerialiser = exports.ConfigurationError = exports.AssertionError = void 0;
const AssertionError_1 = require("./AssertionError");
const ConfigurationError_1 = require("./ConfigurationError");
const ErrorSerialiser_1 = require("./ErrorSerialiser");
const ImplementationPendingError_1 = require("./ImplementationPendingError");
const LogicError_1 = require("./LogicError");
const TestCompromisedError_1 = require("./TestCompromisedError");
const TimeoutExpiredError_1 = require("./TimeoutExpiredError");
const UnknownError_1 = require("./UnknownError");
ErrorSerialiser_1.ErrorSerialiser.registerErrorTypes(AssertionError_1.AssertionError, ConfigurationError_1.ConfigurationError, ImplementationPendingError_1.ImplementationPendingError, LogicError_1.LogicError, TestCompromisedError_1.TestCompromisedError, TimeoutExpiredError_1.TimeoutExpiredError, UnknownError_1.UnknownError);
var AssertionError_2 = require("./AssertionError");
Object.defineProperty(exports, "AssertionError", { enumerable: true, get: function () { return AssertionError_2.AssertionError; } });
var ConfigurationError_2 = require("./ConfigurationError");
Object.defineProperty(exports, "ConfigurationError", { enumerable: true, get: function () { return ConfigurationError_2.ConfigurationError; } });
var ErrorSerialiser_2 = require("./ErrorSerialiser");
Object.defineProperty(exports, "ErrorSerialiser", { enumerable: true, get: function () { return ErrorSerialiser_2.ErrorSerialiser; } });
var ErrorStackParser_1 = require("./ErrorStackParser");
Object.defineProperty(exports, "ErrorStackParser", { enumerable: true, get: function () { return ErrorStackParser_1.ErrorStackParser; } });
var ImplementationPendingError_2 = require("./ImplementationPendingError");
Object.defineProperty(exports, "ImplementationPendingError", { enumerable: true, get: function () { return ImplementationPendingError_2.ImplementationPendingError; } });
var ListItemNotFoundError_1 = require("./ListItemNotFoundError");
Object.defineProperty(exports, "ListItemNotFoundError", { enumerable: true, get: function () { return ListItemNotFoundError_1.ListItemNotFoundError; } });
var LogicError_2 = require("./LogicError");
Object.defineProperty(exports, "LogicError", { enumerable: true, get: function () { return LogicError_2.LogicError; } });
var RuntimeError_1 = require("./RuntimeError");
Object.defineProperty(exports, "RuntimeError", { enumerable: true, get: function () { return RuntimeError_1.RuntimeError; } });
var TestCompromisedError_2 = require("./TestCompromisedError");
Object.defineProperty(exports, "TestCompromisedError", { enumerable: true, get: function () { return TestCompromisedError_2.TestCompromisedError; } });
var TimeoutExpiredError_2 = require("./TimeoutExpiredError");
Object.defineProperty(exports, "TimeoutExpiredError", { enumerable: true, get: function () { return TimeoutExpiredError_2.TimeoutExpiredError; } });
var UnknownError_2 = require("./UnknownError");
Object.defineProperty(exports, "UnknownError", { enumerable: true, get: function () { return UnknownError_2.UnknownError; } });
